src/azure-cli-core/azure/cli/core/command_recommender.py

```diff
@@ -127,93 +180,163 @@ def _set_aladdin_recommendations(self):
         recommendations = []
         if response and response.status_code == HTTPStatus.OK:
             for result in response.json():
-                # parse the response and format the recommendation
-                command, parameters, placeholders = result['command'],\
-                    result['parameters'].split(','),\
-                    result['placeholders'].split('♠')
-                recommendation = 'az {} '.format(command)
-                for parameter, placeholder in zip(parameters, placeholders):
-                    recommendation += '{} {} '.format(parameter, placeholder)
-                recommendations.append(recommendation.strip())
+                # parse the response to get the raw command
+                raw_command = 'az {} '.format(result['command'])
+                for parameter, placeholder in zip(result['parameters'].split(','), result['placeholders'].split('♠')):
+                    raw_command += '{} {}{}'.format(parameter, placeholder, ' ' if placeholder else '')
+
+                # format the recommendation
+                recommendation = {
+                    'command': raw_command.strip(),
+                    'description': result['description'],
+                    'link': result['link']
+                }
+                recommendations.append(recommendation)
 
         self.aladdin_recommendations.extend(recommendations)
 
-    def recommend_a_command(self):
-        """Recommend a command for user when user's command fails.
-        The recommended command will be the best matched one from
-        both the help files and the aladdin recommendations.
+    def provide_recommendations(self):
+        """Provide recommendations when a command fails.
+
+        The recommendations are either from Aladdin service or CLI help examples,
+        which include both commands and reference links along with their descriptions.
+
+        :return: The decorated recommendations
+        :type: list
         """
 
-        def get_sorted_candidates(target_args, candidate_args_list):
-            """Get the sorted candidates by target arguments"""
+        from azure.cli.core.style import Style, get_styled_command
+        from azure.cli.core.parser import OVERVIEW_REFERENCE
+
+        def sort_recommendations(recommendations):
+            """Sort the recommendations by parameter matching.
+
+            The sorting rules below are applied in order:
+                1. Commands starting with the user's input command name are ahead of those don't
+                2. Commands having more matched arguments are ahead of those having less
+                3. Commands having less arguments are ahead of those having more
```

Comment:
Does Aladdin recommendation service have its own ranking rules?
I personally think that the content of relevance ranking should be better implemented by the recommendation server side.

Reply:
Yeah, we discussed with Aladdin team about this. We agreed that before the service could return better ordered recommendtions, we would control the orders in client side.

What's more, the rules will also be applied to example recommendations.